File: 0592-fraction-addition-and-subtraction/0592-fraction-addition-and-subtraction.java
class Solution {
    public String fractionAddition(String expression) {
        expression = expression.replaceAll("-", "\\+-");
        String[] fractions = expression.split("\\+");
        
        List<int[]> fr = new ArrayList();
        
        int commonDenom = 1;
        
        for (String s : fractions) {
            if (s.length() == 0)
                continue;
            
            String[] a = s.split("/");
            int n = Integer.parseInt(a[0]);
            int d = Integer.parseInt(a[1]);
            
            if (commonDenom % d != 0)
                commonDenom *= d;
            fr.add(new int[]{n, d});
        }
        
        int sum = 0;
        for (int[] a : fr) {
            sum += (a[0] * commonDenom / a[1]);
        }

        int gcdSumDenom = gcd(Math.abs(sum), commonDenom);
        sum /= gcdSumDenom;
        commonDenom /= gcdSumDenom;
        
        return "" + sum + "/" + (commonDenom == 0 ? 1 : commonDenom);
    }
    
    public int gcd(int a, int b) {
        while (b != 0) {
            int t = b;
            b = a % b;
            a = t;
        }
        return a;
    }
}

File: 0592-fraction-addition-and-subtraction/0592-fraction-addition-and-subtraction.cpp
class Solution {
public:
    string fractionAddition(string expression) {
        int num = 0;
        int denom = 1;

        int i = 0;
        while (i < expression.size()) {
            int currNum = 0;
            int currDenom = 0;

            bool isNegative = false;

            // check for sign
            if (expression[i] == '-' || expression[i] == '+') {
                if (expression[i] == '-') {
                    isNegative = true;
                }
                // move to next character
                i++;
            }

            // build numerator
            while (isdigit(expression[i])) {
                int val = expression[i] - '0';
                currNum = currNum * 10 + val;
                i++;
            }

            if (isNegative) currNum *= -1;

            // skip divisor
            i++;

            // build denominator
            while (i < expression.size() && isdigit(expression[i])) {
                int val = expression[i] - '0';
                currDenom = currDenom * 10 + val;
                i++;
            }

            // add fractions together using common denominator
            num = num * currDenom + currNum * denom;
            denom = denom * currDenom;
        }

        int gcd = abs(FindGCD(num, denom));

        // reduce fractions
        num /= gcd;
        denom /= gcd;

        return to_string(num) + "/" + to_string(denom);
    }

private:
    int FindGCD(int a, int b) {
        if (a == 0) return b;
        return FindGCD(b % a, a);
    }
};